Mixing/Base/src/PileUp.cc

```diff
@@ -202,24 +203,42 @@ namespace edm {
     input_->doBeginJob();
     if (provider_.get() != nullptr) {
       edm::ServiceRegistry::Operate guard(*serviceToken_);
-      provider_->beginJob(*productRegistry_, iES);
+      GlobalContext globalContext(GlobalContext::Transition::kBeginJob, processContext_.get());
+      provider_->beginJob(*productRegistry_, iES, globalContext);
     }
   }
 
   void PileUp::beginStream(edm::StreamID) {
     auto iID = eventPrincipal_->streamID();  // each producer has its own workermanager, so use default streamid
     streamContext_.reset(new StreamContext(iID, processContext_.get()));
+    streamContext_->setTransition(StreamContext::Transition::kBeginStream);
     if (provider_.get() != nullptr) {
       edm::ServiceRegistry::Operate guard(*serviceToken_);
       provider_->beginStream(iID, *streamContext_);
     }
   }
 
   void PileUp::endStream() {
+    ExceptionCollector exceptionCollector(
+        "Multiple exceptions were thrown while executing PileUp::endStream. An exception message follows for "
+        "each.\n");
+    endStream(exceptionCollector);
+
+    if (exceptionCollector.hasThrown()) {
+      exceptionCollector.rethrow();
+    }
+  }
+
+  void PileUp::endStream(ExceptionCollector& exceptionCollector) {
     if (provider_.get() != nullptr) {
       edm::ServiceRegistry::Operate guard(*serviceToken_);
-      provider_->endStream(streamContext_->streamID(), *streamContext_);
-      provider_->endJob();
+      streamContext_->setTransition(StreamContext::Transition::kEndStream);
+      provider_->endStream(streamContext_->streamID(), *streamContext_, exceptionCollector);
+      // This is kind of strange, end of job running as part of endStream multiple times...
+      // For the moment, I'm leaving this as is but maybe we should think about this...
+      // I think nothing uses this code anymore anyway...
```

Comment:
We should probably open an issue about this question (or, an issue would have higher chances of being addressed at some point).

Reply:
I added a new Issue #45506.